src/server/scripts/Kalimdor/HallsOfOrigination/boss_isiset.cpp

```diff
@@ -0,0 +1,1075 @@
+            events.ScheduleEvent(EVENT_STARRY_SKY_ADD_VISUAL, 800); // Sniffed timers
+            me->DespawnOrUnsummon(3600);
+        }
+
+        void UpdateAI(uint32 diff) override
+        {
+            events.Update(diff);
+
+            if (me->HasUnitState(UNIT_STATE_CASTING))
+                return;
+
+            while (uint32 eventId = events.ExecuteEvent())
+            {
+                switch (eventId)
+                {
+                    case EVENT_STARRY_SKY_ADD_VISUAL:
+                        DoCastSelf(SPELL_STARRY_SKY_VISUAL);
+                        break;
+                    default:
+                        break;
+                }
+            }
+        }
+
+    private:
+        EventMap events;
+    };
+
+    CreatureAI* GetAI(Creature* creature) const override
+    {
+        return GetHallsOfOriginationAI<npc_starry_skyAI>(creature);
+    }
+};
+
+// 39720 - Astral Rain
+// 39721 - Celestial Call
+// 39722 - Veil of Sky
+class npc_isiset_mirror_image : public CreatureScript
+{
+public:
+    npc_isiset_mirror_image() : CreatureScript("npc_isiset_mirror_image") { }
+
+    struct npc_isiset_mirror_imageAI : public ScriptedAI
+    {
+        npc_isiset_mirror_imageAI(Creature* creature) : ScriptedAI(creature), _instance(creature->GetInstanceScript())
+        {
+            me->SetReactState(REACT_PASSIVE);
+            _instance->SendEncounterUnit(ENCOUNTER_FRAME_ENGAGE, me, 1);
+
+            switch (me->GetEntry()) {
+                case NPC_ASTRAL_RAIN:
+                    _events.ScheduleEvent(EVENT_IMAGE_ASTRAL_RAIN_ABILITY, Seconds(3), Seconds(5));
+                    DoCastSelf(SPELL_STARFALL_STATE_VISUAL);
+                    break;
+                case NPC_CELESTIAL_CALL:
+                    _events.ScheduleEvent(EVENT_IMAGE_ARCANE_BARRAGE_ABILITY, Seconds(6));
+                    DoCastSelf(SPELL_ADDS_STATE_VISUAL);
+                    break;
+                case NPC_VEIL_OF_SKY:
+                    _events.ScheduleEvent(EVENT_IMAGE_VEIL_OF_SKY_ABILITY, Seconds(8), Seconds(9));
+                    //DoCastSelf(SPELL_MANA_SHIELD_STATE_VISUAL); // Disabled in sniffs/on retail, too shiny. 
+                    break;
+            }
+        }
+
+        void DamageTaken(Unit* /*attacker*/, uint32 &damage) override
+        {
+            // Because only one image explodes, others become passive
+            if (me->GetReactState() != REACT_AGGRESSIVE || me->GetHealth() > damage)
+                return;
+
+            me->SetReactState(REACT_PASSIVE);
+            damage = me->GetHealth() - 1;
+            DoCastSelf(SPELL_IMAGE_EXPLOSION);
+        }
+
+        void DoAction(int32 action) override
+        {
+            switch (action)
+            {
+                case ACTION_IMAGES_SET_AGGRESSIVE:
+                    me->SetReactState(REACT_AGGRESSIVE);
+                    me->SetInCombatWithZone();
+                    AttackStart(SelectTarget(SELECT_TARGET_RANDOM));
```

Comment:
Check if generic AttackStart has protection to null.
Because SelectTarget(SELECT_TARGET_RANDOM) can return null xd